File: Rust/fractionalization.rs
// © 2019 Aaron Sami Abassi
// Licensed under the Academic Free License version 3.0
mod fractionalization {
    use num::Num;
    use num::Integer;
    pub type Fractional<Natural> = (Natural, Natural);
    pub type ArithmeticalAbstract<Natural> = fn(Fractional<Natural>, Fractional<Natural>) -> Fractional<Natural>;
    pub type RelationalAbstract<Natural> = fn(Fractional<Natural>, Fractional<Natural>) -> bool;
    pub struct Arithmetical<Natural> {
        pub add: ArithmeticalAbstract<Natural>,
        pub subtract: ArithmeticalAbstract<Natural>,
        pub multiply: ArithmeticalAbstract<Natural>,
        pub divide: ArithmeticalAbstract<Natural>
    }
    pub struct Relational<Natural> {
        pub lesser: RelationalAbstract<Natural>,
        pub greater: RelationalAbstract<Natural>,
        pub equal: RelationalAbstract<Natural>,
        pub not_greater: RelationalAbstract<Natural>,
        pub not_lesser: RelationalAbstract<Natural>,
        pub not_equal: RelationalAbstract<Natural>
    }
    pub struct Operational<Natural> {
        pub arithmetic: Arithmetical<Natural>,
        pub relation: Relational<Natural>
    }
    pub fn reduce<Natural>(fraction: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Num + Integer + Copy {
        if fraction.0 == Natural::zero() {
            return (Natural::zero(), Natural::one());
        } else {
            let divisor = fraction.0.gcd(&fraction.1);
            return (fraction.0 / divisor, fraction.1 / divisor);
        }
    }
    pub fn fast_add<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        return (base.0 * relative.1 + relative.0 * base.1, base.1 * relative.1);
    }
    pub fn fast_subtract<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        return (base.0 * relative.1 - relative.0 * base.1, base.1 * relative.1);
    }
    pub fn fast_multiply<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        return (base.0 * relative.0, base.1 * relative.1);
    }
    pub fn fast_divide<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        return (base.0 * relative.1, base.1 * relative.0);
    }
    pub fn reducing_add<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        let multiple = base.1.lcm(&relative.1);
        return reduce((base.0 * (multiple / base.1) + relative.0 * (multiple / relative.1), multiple));
    }
    pub fn reducing_subtract<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        let multiple = base.1.lcm(&relative.1);
        return reduce((base.0 * (multiple / base.1) - relative.0 * (multiple / relative.1), multiple));
    }
    pub fn reducing_multiply<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        let bnrd_divisor: Natural;
        let bdrn_divisor: Natural;
        if base.0 > Natural::zero() {
            bnrd_divisor = base.0.gcd(&relative.1);
        } else {
            bnrd_divisor = Natural::one();
        }
        if relative.0 > Natural::zero() {
            bdrn_divisor = base.1.gcd(&relative.0);
        } else {
            bdrn_divisor = Natural::one();
        }
        return ((base.0 / bnrd_divisor) * (relative.0 / bdrn_divisor), (base.1 / bdrn_divisor) * (relative.1 / bnrd_divisor));
    }
    pub fn reducing_divide<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> Fractional<Natural> 
    where Natural: Integer + Copy {
        let bnrn_divisor = base.0.gcd(&relative.0);
        let bdrd_divisor = base.1.gcd(&relative.1);
        return ((base.0 / bnrn_divisor) * (relative.1 / bdrd_divisor), (base.1 / bdrd_divisor) * (relative.0 / bnrn_divisor));
    }
    pub fn fast_lesser<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer {
        return base.0 * relative.1 < relative.0 * base.1;
    }
    pub fn fast_greater<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer {
        return base.0 * relative.1 > relative.0 * base.1;
    }
    pub fn fast_equal<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer {
        return base.0 * relative.1 == relative.0 * base.1;
    }
    pub fn fast_not_greater<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer {
        return base.0 * relative.1 <= relative.0 * base.1;
    }
    pub fn fast_not_lesser<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer {
        return base.0 * relative.1 >= relative.0 * base.1;
    }
    pub fn fast_not_equal<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer {
        return base.0 * relative.1 != relative.0 * base.1;
    }
    pub fn reducing_lesser<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer + Copy {
        let factor = base.1.lcm(&relative.1);
        return base.0 * (factor / base.1) < relative.0 * (factor / relative.1);
    }
    pub fn reducing_greater<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer + Copy {
        let factor = base.1.lcm(&relative.1);
        return base.0 * (factor / base.1) > relative.0 * (factor / relative.1);
    }
    pub fn reducing_equal<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer + Copy {
        let factor = base.1.lcm(&relative.1);
        return base.0 * (factor / base.1) == relative.0 * (factor / relative.1);
    }
    pub fn reducing_not_greater<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer + Copy {
        let factor = base.1.lcm(&relative.1);
        return base.0 * (factor / base.1) <= relative.0 * (factor / relative.1);
    }
    pub fn reducing_not_lesser<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer + Copy {
        let factor = base.1.lcm(&relative.1);
        return base.0 * (factor / base.1) >= relative.0 * (factor / relative.1);
    }
    pub fn reducing_not_equal<Natural>(base: Fractional<Natural>, relative: Fractional<Natural>) -> bool 
    where Natural: Integer + Copy {
        let factor = base.1.lcm(&relative.1);
        return base.0 * (factor / base.1) != relative.0 * (factor / relative.1);
    }
    pub fn fast_arithmetic<Natural>() -> Arithmetical<Natural> 
    where Natural: Integer + Copy {
        let arithmetic = Arithmetical::<Natural> {
            add: fast_add::<Natural>,
            subtract: fast_subtract::<Natural>,
            multiply: fast_multiply::<Natural>,
            divide: fast_divide::<Natural>
        };
        return arithmetic;
    }
    pub fn reducing_arithmetic<Natural>() -> Arithmetical<Natural> 
    where Natural: Integer + Copy {
        let arithmetic = Arithmetical::<Natural> {
            add: reducing_add::<Natural>,
            subtract: reducing_subtract::<Natural>,
            multiply: reducing_multiply::<Natural>,
            divide: reducing_divide::<Natural>
        };
        return arithmetic;
    }
    pub fn fast_relation<Natural>() -> Relational<Natural> 
    where Natural: Integer {
        let relation = Relational::<Natural> {
            lesser: fast_lesser::<Natural>,
            greater: fast_greater::<Natural>,
            equal: fast_equal::<Natural>,
            not_greater: fast_not_greater::<Natural>,
            not_lesser: fast_not_lesser::<Natural>,
            not_equal: fast_not_equal::<Natural>
        };
        return relation;
    }
    pub fn reducing_relation<Natural>() -> Relational<Natural> 
    where Natural: Integer + Copy {
        let relation = Relational::<Natural> {
            lesser: reducing_lesser::<Natural>,
            greater: reducing_greater::<Natural>,
            equal: reducing_equal::<Natural>,
            not_greater: reducing_not_greater::<Natural>,
            not_lesser: reducing_not_lesser::<Natural>,
            not_equal: reducing_not_equal::<Natural>
        };
        return relation;
    }
    pub fn fast_operation<Natural>() -> Operational<Natural> 
    where Natural: Integer + Copy {
        let operation = Operational::<Natural> {
            arithmetic: fast_arithmetic::<Natural>(),
            relation: fast_relation::<Natural>()
        };
        return operation;
    }
    pub fn reducing_operation<Natural>() -> Operational<Natural> 
    where Natural: Integer + Copy {
        let operation = Operational::<Natural> {
            arithmetic: reducing_arithmetic::<Natural>(),
            relation: reducing_relation::<Natural>()
        };
        return operation;
    }
}

File: Visual Basic/Fractionalization.vb
﻿' © 2018 Aaron Sami Abassi
' Licensed under the Academic Free License version 3.0
Module Fractionalization

    Public Delegate Function ArithmeticalAbstract(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional

    Public Delegate Function RelationalAbstract(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean

    Public Structure Fractional

        Dim Numerator, Denominator As UInteger

        Sub New(ByRef Numerator As UInteger, Optional ByRef Denominator As UInteger = 1)
            Me.Numerator = Numerator
            Me.Denominator = Denominator
        End Sub

    End Structure

    Public Structure Arithmetical

        ReadOnly Add, Subtract, Multiply, Divide As ArithmeticalAbstract

        Sub New(ByRef Add As ArithmeticalAbstract, ByRef Subtract As ArithmeticalAbstract, ByRef Multiply As ArithmeticalAbstract, ByRef Divide As ArithmeticalAbstract)
            Me.Add = Add
            Me.Subtract = Subtract
            Me.Multiply = Multiply
            Me.Divide = Divide
        End Sub

    End Structure

    Public Structure Relational

        ReadOnly Lesser, Greater, Equal, NotGreater, NotLesser, NotEqual As RelationalAbstract

        Sub New(ByRef Lesser As RelationalAbstract, ByRef Greater As RelationalAbstract, ByRef Equal As RelationalAbstract, ByRef NotGreater As RelationalAbstract, ByRef NotLesser As RelationalAbstract, ByRef NotEqual As RelationalAbstract)
            Me.Lesser = Lesser
            Me.Greater = Greater
            Me.Equal = Equal
            Me.NotGreater = NotGreater
            Me.NotLesser = NotLesser
            Me.NotEqual = NotEqual
        End Sub

    End Structure

    Public Structure Operational

        ReadOnly Arithmetic As Arithmetical

        ReadOnly Relation As Relational

        Sub New(ByRef Arithmetic As Arithmetical, ByRef Relation As Relational)
            Me.Arithmetic = Arithmetic
            Me.Relation = Relation
        End Sub

    End Structure

    Public ReadOnly Zero As New Fractional(0, 1)

    Public Function GreatestCommonDivisor(A As UInteger, B As UInteger) As UInteger
        ' Eclidean Algorithm
        Dim Dividend, Divisor, Remainder As UInteger
        If A > B Then
            Dividend = A
            Divisor = B
        Else
            Dividend = B
            Divisor = A
        End If
        While (Remainder = Dividend Mod Divisor) > 0
            Dividend = Divisor
            Divisor = Remainder
        End While
        Return Divisor
    End Function

    Public Function LeastCommonMultiple(A As UInteger, B As UInteger) As UInteger
        Dim Divisor As UInteger
        Divisor = GreatestCommonDivisor(A, B)
        If A > B Then
            Return A / Divisor * B
        End If
        Return B / Divisor * A
    End Function

    Public Function Reduce(ByRef Fraction As Fractional) As Fractional
        Dim Divisor As UInteger
        Dim Result As Fractional
        If Fraction.Numerator = 0 Then
            Result.Numerator = 0
            Result.Denominator = 1
        Else
            Divisor = GreatestCommonDivisor(Fraction.Numerator, Fraction.Denominator)
            Result.Numerator = Fraction.Numerator / Divisor
            Result.Denominator = Fraction.Denominator / Divisor
        End If
        Return Result
    End Function

    Public Function FastAdd(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Denominator + Relative.Numerator * Base.Denominator
        Result.Denominator = Base.Denominator * Relative.Denominator
        Return Result
    End Function

    Public Function FastSubtract(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Denominator - Relative.Numerator * Base.Denominator
        Result.Denominator = Base.Denominator * Relative.Denominator
        Return Result
    End Function

    Public Function FastMultiply(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Numerator
        Result.Denominator = Base.Denominator * Relative.Denominator
        Return Result
    End Function

    Public Function FastDivide(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Denominator
        Result.Denominator = Base.Denominator * Relative.Numerator
        Return Result
    End Function

    Public Function ReducingAdd(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Denominator = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Result.Numerator = Base.Numerator * (Result.Denominator / Base.Denominator)
        Result.Numerator += Relative.Numerator * (Result.Denominator / Relative.Denominator)
        Return Reduce(Result)
    End Function

    Public Function ReducingSubtract(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Denominator = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Result.Numerator = Base.Numerator * (Result.Denominator / Base.Denominator)
        Result.Numerator -= Relative.Numerator * (Result.Denominator / Relative.Denominator)
        Return Reduce(Result)
    End Function

    Public Function ReducingMultiply(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim BNRDDivisor, BDRNDivisor As UInteger
        Dim Result As Fractional
        If Base.Numerator > 0 Then
            BNRDDivisor = GreatestCommonDivisor(Base.Numerator, Relative.Denominator)
        Else
            BNRDDivisor = 1
        End If
        If Relative.Numerator > 0 Then
            BDRNDivisor = GreatestCommonDivisor(Base.Denominator, Relative.Numerator)
        Else
            BDRNDivisor = 1
        End If
        Result.Numerator = (Base.Numerator / BNRDDivisor) * (Relative.Numerator / BDRNDivisor)
        Result.Denominator = (Base.Denominator / BDRNDivisor) * (Relative.Denominator / BNRDDivisor)
        Return Result
    End Function

    Public Function ReducingDivide(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim BNRNDivisor, BDRDDivisor As UInteger
        Dim Result As Fractional
        BNRNDivisor = GreatestCommonDivisor(Base.Numerator, Relative.Numerator)
        BDRDDivisor = GreatestCommonDivisor(Base.Denominator, Relative.Denominator)
        Result.Numerator = (Base.Numerator / BNRNDivisor) * (Relative.Denominator / BDRDDivisor)
        Result.Denominator = (Base.Denominator / BDRDDivisor) * (Relative.Numerator / BNRNDivisor)
        Return Result
    End Function

    Public Function FastLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator < Relative.Numerator * Base.Denominator
    End Function

    Public Function FastGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator > Relative.Numerator * Base.Denominator
    End Function

    Public Function FastEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator = Relative.Numerator * Base.Denominator
    End Function

    Public Function FastNotGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator <= Relative.Numerator * Base.Denominator
    End Function

    Public Function FastNotLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator >= Relative.Numerator * Base.Denominator
    End Function

    Public Function FastNotEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator <> Relative.Numerator * Base.Denominator
    End Function

    Public Function ReducingLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) < Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) > Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) = Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingNotGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) <= Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingNotLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) >= Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingNotEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) <> Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public ReadOnly FastArithmetic As New Arithmetical(AddressOf FastAdd, AddressOf FastSubtract, AddressOf FastMultiply, AddressOf FastDivide)

    Public ReadOnly ReducingArithmetic As New Arithmetical(AddressOf ReducingAdd, AddressOf ReducingSubtract, AddressOf ReducingMultiply, AddressOf ReducingDivide)

    Public ReadOnly FastRelation As New Relational(AddressOf FastLesser, AddressOf FastGreater, AddressOf FastEqual, AddressOf FastNotGreater, AddressOf FastNotLesser, AddressOf FastNotEqual)

    Public ReadOnly ReducingRelation As New Relational(AddressOf ReducingLesser, AddressOf ReducingGreater, AddressOf ReducingEqual, AddressOf ReducingNotGreater, AddressOf ReducingNotLesser, AddressOf ReducingNotEqual)

    Public ReadOnly FastOperation As New Operational(FastArithmetic, FastRelation)

    Public ReadOnly ReducingOperation As New Operational(ReducingArithmetic, ReducingRelation)

End Module
